tests/wasm_tests/tic_tac_toe_tests.cpp

```diff
@@ -0,0 +1,211 @@
+           ("challenger", "player2")
+           ("host", "player1")
+   );
+
+   chain.produce_blocks();
+
+   auto mvt = mutable_variant_object()
+           ("row", 1)
+           ("column", 1);
+   chain.push_action(N(tic.tac.toe), N(move), N(player1), mutable_variant_object()
+           ("challenger", "player2")
+           ("host", "player1")
+           ("by", "player1")
+           ("mvt", mvt)
+   );
+
+   mvt = mutable_variant_object()
+           ("row", 0)
+           ("column", 1);
+   BOOST_CHECK_EXCEPTION(chain.push_action(N(tic.tac.toe), N(move), N(player1), mutable_variant_object()
+         ("challenger", "player2")
+         ("host", "player1")
+         ("by", "player1")
+         ("mvt", mvt)
+      ), transaction_exception, assert_message_contains("not your turn"));
+
+   mvt = mutable_variant_object()
+           ("row", 1)
+           ("column", 1);
+   BOOST_CHECK_EXCEPTION(chain.push_action(N(tic.tac.toe), N(move), N(player2), mutable_variant_object()
+         ("challenger", "player2")
+         ("host", "player1")
+         ("by", "player2")
+         ("mvt", mvt)
+      ), transaction_exception, assert_message_contains("not a valid movement"));
+
+   mvt = mutable_variant_object()
+           ("row", 0)
+           ("column", 1);
+   chain.push_action(N(tic.tac.toe), N(move), N(player2), mutable_variant_object()
+           ("challenger", "player2")
+           ("host", "player1")
+           ("by", "player2")
+           ("mvt", mvt)
+   );
+
+   mvt = mutable_variant_object()
+           ("row", 0)
+           ("column", 0 );
+   chain.push_action(N(tic.tac.toe), N(move), N(player1), mutable_variant_object()
+           ("challenger", "player2")
+           ("host", "player1")
+           ("by", "player1")
+           ("mvt", mvt)
+   );
+
+   mvt = mutable_variant_object()
+           ("row", 0)
+           ("column", 2);
+   chain.push_action(N(tic.tac.toe), N(move), N(player2), mutable_variant_object()
+           ("challenger", "player2")
+           ("host", "player1")
+           ("by", "player2")
+           ("mvt", mvt)
+   );
+
+   mvt = mutable_variant_object()
+           ("row", 2)
+           ("column", 2);
+   chain.push_action(N(tic.tac.toe), N(move), N(player1), mutable_variant_object()
+           ("challenger", "player2")
+           ("host", "player1")
+           ("by", "player1")
+           ("mvt", mvt)
+   );
+
+   mvt = mutable_variant_object()
+           ("row", 2)
+           ("column", 0);
+   BOOST_CHECK_EXCEPTION(chain.push_action(N(tic.tac.toe), N(move), N(player2), mutable_variant_object()
+         ("challenger", "player2")
+         ("host", "player1")
+         ("by", "player2")
+         ("mvt", mvt)
+      ), transaction_exception, assert_message_contains("the game has ended"));
+
+   chain.push_action(N(tic.tac.toe), N(close), N(player1), mutable_variant_object()
+           ("challenger", "player2")
+           ("host", "player1")
+   );
```

Comment:
Can you add step to verify winner.

Reply:
Yes, also added Tester methods to access contract table data.